cmd/thanos/config.go

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package main
+
+import (
+	"net/url"
+	"time"
+
+	"github.com/bwplotka/flagarize"
+	"github.com/prometheus/common/model"
+)
+
+type grpcConfig struct {
+	BindAddress    string         `flagarize:"name=Grpc-address|help=Listen ip:port address for gRPC endpoints (StoreAPI). Must be routable from other components.|default=0.0.0.0:10901"`
```

Comment:
I'm not a huge fan of the struct tags in this way. We're throwing away all the type safety and language characteristics that we could previously use with kingpin. Why can't we capture the information kept in the struct tags here in real structs?

Reply:
@brancz Thats the approach I took in the initial commit https://github.com/thanos-io/thanos/pull/2267/commits/2562b321fbd8d7dd6073f65299272c67d824c1af

I just integrated the library after the first review as suggested but am happy to revert if thats the consensus

Reply:
I think I would personally prefer that approach.

Reply:
@brancz - I can let yourself and @bwplotka battle that one out in that case :D

Reply:
I can see lack of type safety, but the advantage is huge - we have finally a way to tie struct filed directly, in the same readable place to a flag. In the same way we can do at some point with `yaml` or json parsing if we choose to have config based configuration. 

Also we literally don't need type safety here, because we always check on our CI the issue on the very start by doing `thanos --help` on every command. (make docs). So overall it does not matter if the error on registering flags is in compile-time or runtime. That's why, while I love type safety, here I am able to sacrifice is for a very good reason.

Reply:
If you still disagree it would be nice to have opinions from other maintainers as well e.g: @squat @kakkoyun @povilasv @GiedriusS

Reply:
Overall yea, let's discuss and ensure we all agree here. I can see this being "bit surprising" approach (:

Reply:
I'm also not a big fan of struct tags, but I think we can work around it by having good linter in place, which would complain if it can't parse the tag and explain whats wrong.

Reply:
_My very personal perspective:_

To my understanding, the plan is to introduce a linter to do the compiler job just because we want to specify CLI flags as struct tags instead of code. Why do we want it? Because "we have finally a way to tie struct filed directly, in the same readable place to a flag".

As you probably know, in Cortex we basically achieve the same result attaching a `RegisterFlags()` function to each config struct and we recursively call `RegisterFlags()` on sub-config structs. Despite there's more code involved, I personally find it a safer approach (and avoids to have to spend further time writing a linter).

Example:
https://github.com/cortexproject/cortex/blob/9245c3120d7b16117aaffcb383a6347d002475d1/pkg/compactor/compactor.go#L52-L61

Reply:
Awesome, thanks for the feedback. Looks like flagarize is not the best option here. I can totally agree with your points. As always there is some power in type safety and more important: autocompletion. (: 

@PhilipGough sorry for taking this back and forth and thanks for giving us the opportunity to see how flagarize would work in Thanos (:

To me looks like we settled on just:

```go
type ...Config struct {
    httpAddress string
}
c := ...Config{}
cmd.Flag( "name", "help").Default().StringVar(&c.httpAddress)
``` 

It's almost similar to what Cortex has except one thing: I am pretty sure we should never define flags outside main package.

BTW,  I only wished @brancz would have the same opinion for other languages and patterns... ;p

Reply:
I've sort of like flagarize, as we would have only one place where we define the config, instead of flags & struct.
The linter thing doesn't have to happen now, so the PR & approach to me looks good.

Reply:
I actually quite like the pattern in Cortex.

Reply:
Let's discuss it more I feel like we have 3 options already (: 

1. Flagarize
2. Struct with typed flag registration in main
3.  Struct with typed flag registration in each component.

I believe flag should never be part of package. We can then limit the surfaces we should care  for each package. If suddenly add flags we add much more things we need to care for in terms of complexity, compatibility etc on top of what we already have with metrics and logs... Imagine e.g reloader package is used in Thanos and X projects (e.g part of prometheus-operator is using it). If we would add flag registration to reloader flag we introduce huge pain in renaming, what should be default should be hidden or not? Quite a pain ): 

So for me no 2 wins, we could think about flagarize more, but looks like 2 is good step in this direction we all are comfortable to take IMO (:

Reply:
I'm happy with that as a first step.

Reply:
> I believe flag should never be part of package

That's what happen with flagarize too. Being coupled with the config struct, whenever you move it to a package flags definition moves along it.

> 2. Struct with typed flag registration in main

Even if this option misses the point of "we have finally a way to tie struct filed directly, in the same readable place to a flag", I still see it valuable to move config to `struct` 👍

Reply:
> That's what happen with flagarize too

Well that's the point we should never use even flagarize outside of main (:

Reply:
I have a bit different opinion, one thing having multiple structs / config options in packages gives us that you can add build tags which could be compiled for embedded stuff. 

For example I'm using s3 and I don't need all that azure stuff and it's dependencies. This would make binary size way smaller, etc. 

One example of this pattern is node exporter actually does command line flags in the different packages and it's fine:

https://github.com/prometheus/node_exporter/blob/master/collector/systemd_linux.go#L41-L48

https://github.com/prometheus/node_exporter/blob/master/collector/filesystem_linux.go#L38-L41

Reply:
Interesting. @povilasv you are right but we can already do that as we hide all the plugins with the config pathOrContent yaml (: